src/screencast/screencast.ts

```diff
@@ -0,0 +1,193 @@
+            });
+        }
+    }
+
+    private updateHistory(): void {
+        this.cdpConnection.sendMessageToBackend('Page.getNavigationHistory', {}, (result) => {
+            const {currentIndex, entries} = result;
+            this.history = entries;
+            this.historyIndex = currentIndex;
+            this.backButton.disabled = this.historyIndex < 1;
+            this.forwardButton.disabled = this.historyIndex >= this.history.length - 1;
+            this.urlInput.value = this.history[this.historyIndex].url;
+        });
+    }
+
+    private updateEmulation(): void {
+        const params = {
+            width: this.width || this.screencastWrapper.offsetWidth,
+            height: this.height || this.screencastWrapper.offsetHeight,
+            deviceScaleFactor: 0,
+            mobile: this.isDeviceTouch()
+        }
+        this.cdpConnection.sendMessageToBackend('Emulation.setDeviceMetricsOverride', params);
+        this.updateScreencast();
+    }
+
+    private isDeviceTouch(){
+        const selectedOption = this.deviceSelect[this.deviceSelect.selectedIndex];
+        return selectedOption.getAttribute('touch') === "true" 
+            || selectedOption.getAttribute('mobile') === "true";
+    }
+
+    private updateScreencast(): void {
+        const screencastParams = {
+            format: 'png',
+            quality: 100,
+            maxWidth: this.width || this.screencastWrapper.offsetWidth,
+            maxHeight: this.height || this.screencastWrapper.offsetHeight
+        };
+        this.cdpConnection.sendMessageToBackend('Page.startScreencast', screencastParams);
+    }
+
+    private onBackClick(): void {
+        if (this.historyIndex > 0) {
+            const entryId = this.history[this.historyIndex - 1].id;
+            this.cdpConnection.sendMessageToBackend('Page.navigateToHistoryEntry', {entryId})
+        }
+    }
+
+    private onDomContentEventFired(): void {
+        setTimeout(() => this.updateEmulation(), 100);
+    }
+
+    private onForwardClick(): void {
+        if (this.historyIndex < this.history.length - 1) {
+            const entryId = this.history[this.historyIndex + 1].id;
+            this.cdpConnection.sendMessageToBackend('Page.navigateToHistoryEntry', {entryId})
+        }
+    }
+
+    private onFrameNavigated({frame}: any): void {
+        if (!frame.parentId) {
+            this.updateHistory();
+            this.updateEmulation();
+        }
+    }
+
+    private onReloadClick(): void {
+        this.cdpConnection.sendMessageToBackend('Page.reload', {});
+    }
+
+    private onRotateClick(): void {
+        const width = this.height;
+        const height = this.width;
+        this.width = width;
+        this.height = height;
+        this.updateEmulation();
+    }
+
+    private onUrlKeyDown(event: KeyboardEvent): void {
+        let url = this.urlInput.value;
+        if (event.key === 'Enter' && url) {
+            if(!url.startsWith('http')) {
```

Comment:
Probably should also allow the `file:` protocol (though I doubt it's that likely to be directly typed in).

Reply:
done.